### PR TITLE
[WIP] Start redesign

### PR DESCRIPTION
**Pull request by LightDiscord**
Start redesign of backend using cool things like decorators :3

### CODE REVIEW

File: src/resolvers/mutations/animes.js [hunk 1]
```diff
@@ -1,87 +1,91 @@
...
-			.then(() => ({ anime, ...season }))
-	);
+	if (cover) anime.cover = cover;
+	if (background) anime.background = background;
+
+	anime.edit_date = time;
+	return anime;
 }
+
+export const { updateAnime, addAnime, addSeason } = new Animes();
```
Comment (DeltaEvo) [inline src/resolvers/mutations/animes.js]: Awww, i don't like this hack, decorator is not the solution I think

File: src/resolvers/mutations/animes.js [hunk 2]
```diff
@@ -1,87 +1,91 @@
...
 			),
+
 		anime.background &&
 			anime.background.then(background =>
-				storage.save(
-					`${id}_background`,
-					background,
-					sharp().jpeg({ progressive: true, quality: 100, optimiseScans: true })
-				)
+				storage.save(`${id}_background`, background, sharp().jpeg(options))
```
Comment (DeltaEvo) [inline src/resolvers/mutations/animes.js]: Allowing `storage.save` to take a promise as image argument could clear this messy code

File: src/resolvers/mutations/animes.js [hunk 3]
```diff
@@ -1,87 +1,91 @@
...
+			.insertOne({
+				...anime,
+				medias: [],
+				seasons: []
+			})
+			.then(({ insertedId }) => insertedId);
+	}
+
+	@requireGroup(ADMIN)
+	async addSeason(root, { anime, season }, context) {
```
Comment (DeltaEvo) [inline src/resolvers/mutations/animes.js]: `async` here  is not useful

**Comment by LightDiscord** [comment]
Redesign of mutations to review @DeltaEvo

File: src/resolvers/mutations/authors.js [hunk 4]
```diff
@@ -1,40 +1,39 @@
...
-				.collection("authors")
-				.updateOne({ _id: new ObjectID(id) }, { $set: author })
-				.then(() => id);
-		});
+export async function updateAuthor(root, { id, author }, context) {
+	await needGroup(context, ADMIN);
+
+	const picture =
+		author.picture && (await context.storage.save(id, author.picture));
+	author.picture = picture;
```
Comment (DeltaEvo) [inline src/resolvers/mutations/authors.js]:
```javascript
if (author.picture)
    author.picture = await context.storage.save(id, author.picture);
```

File: src/resolvers/mutations/notifications.js [hunk 5]
```diff
@@ -1,13 +1,12 @@
...
+	const { value } = await context.db
 		.collection("notifications")
-		.findOneAndDelete({ _id: new ObjectID(notif) })
-		.then(
-			({ value }) =>
-				!value ? { error: "This notification does not exist" } : { error: null }
-		);
+		.findOneAndDelete({ _id: new ObjectID(notif) });
+
+	return { error: !value ? "This notification does not exist" : null };
```
Comment (DeltaEvo) [inline src/resolvers/mutations/notifications.js]: Don't invert condition

File: src/resolvers/mutations/slider.js [hunk 6]
```diff
@@ -1,52 +1,48 @@
...
-					context.db
-						.collection("slider")
-						.insertOne(slide)
-						.then(({ insertedId }) => ({ id: insertedId, ...slide }))
-				]).then(([, res]) => res);
-			})
-	);
+	const image = slide.image && (await context.storage.save(id, slide.image));
+
+	if (image) slide.image = image;
```
Comment (DeltaEvo) [inline src/resolvers/mutations/slider.js]:
```javascript
if (slide.image)
    slide.image = await context.storage.save(id, slide.image)
```

File: src/resolvers/mutations/slider.js [hunk 7]
```diff
@@ -1,52 +1,48 @@
...
-					.findOneAndUpdate({ _id: new ObjectID(id) }, { $set: slide })
-					.then(({ value }) => ({ id, ...value, ...slide }));
-			})
-	);
+export async function editSlide(root, { id, slide }, context) {
+	await needGroup(context, ADMIN);
+
+	const image = slide.image && (await context.storage.save(id, slide.image));
+
+	if (image) slide.image = image;
```
Comment (DeltaEvo) [inline src/resolvers/mutations/slider.js]: ^

File: src/resolvers/mutations/users.js [hunk 8]
```diff
@@ -14,106 +14,105 @@ const IMAGE_MIME_TYPES = [
...
+	file = await file;
+
+	if (!IMAGE_MIME_TYPES.includes(file.mimetype)) {
+		return { error: `MimeType ${file.mimetype} is not and image Mime Type` };
+	}
+
+	const user = await context.user;
+	const url = await context.storage.save(user.id, file);
+	user.avatar = url;
+	user.save();
```
Comment (DeltaEvo) [inline src/resolvers/mutations/users.js]:
```javascript
await user.save();
```

File: src/resolvers/mutations/users.js [hunk 9]
```diff
@@ -14,106 +14,105 @@ const IMAGE_MIME_TYPES = [
...
+			{ $addToSet: { friends: user.id } },
+			{ upsert: true }
+		);
+
+	user.friends.push(new ObjectID(other));
+
+	newFriendEvent({ user: user.id, friend: new ObjectID(other) }, context);
+	newFriendEvent({ user: new ObjectID(other), friend: user.id }, context);
+
+	return { error: u.save() ? "nothing to save" : null };
```
Comment (DeltaEvo) [inline src/resolvers/mutations/users.js]:
```javascript
return { error: (await u.save()) ? "nothing to save" : null };
```

File: src/storage/Storage.js [hunk 10]
```diff
@@ -1,11 +1,13 @@
 export default class Storage {
 	register(app) {}
 
-	save(filename, { mimetype, stream }, ...transforms) {
-		return this.write(
-			filename + "",
-			mimetype,
-			transforms.reduce((c, fn) => c.pipe(fn), stream)
+	save(filename, file, ...transforms) {
+		return file.then(({ mimetype, stream }) =>
```
Comment (DeltaEvo) [inline src/storage/Storage.js]:
```javascript
return Promise.resolve(file).then(({ mimetype, stream }) =>
```

File: src/resolvers/mutations/animes.js [hunk 11]
```diff
@@ -2,86 +2,79 @@ import { toId, now, needGroup, ADMIN } from "../util/index";
...
-		transformAnime(id, anime, now(), context.storage).then(anime =>
-			context.db
-				.collection("animes")
-				.updateOne({ _id: id }, { $set: anime })
-				.then(() => id)
-		)
-	);
+export async function updateAnime(root, { id, anime }, context) {
+	await needGroup(context, ADMIN);
+	anime = await transformAnime(id, anime, now(), context.storage);
```
Comment (DeltaEvo) [inline src/resolvers/mutations/animes.js]: You don't need to reassign `anime`

File: src/resolvers/mutations/animes.js [hunk 12]
```diff
@@ -2,86 +2,79 @@ import { toId, now, needGroup, ADMIN } from "../util/index";
...
-					sharp().jpeg({ progressive: true, quality: 100, optimiseScans: true })
-				)
-			)
-	]).then(([cover, background]) => {
-		if (cover) anime.cover = cover;
-		if (background) anime.background = background;
-		anime.edit_date = time;
-		return anime;
-	});
+	anime = await transformAnime(id, anime, time, context.storage);
```
Comment (DeltaEvo) [inline src/resolvers/mutations/animes.js]: ^

File: src/resolvers/mutations/authors.js [hunk 13]
```diff
@@ -1,40 +1,38 @@
...
-		.then(picture => {
-			author.picture = picture;
-			author._id = id;
-			return;
-			context.db
-				.collection("authors")
-				.insertOne(author)
-				.then(({ insertedId }) => insertedId);
-		});
+	const picture = author.picture && (await context.storage.save(id, picture));
```
Comment (DeltaEvo) [inline src/resolvers/mutations/authors.js]:
```javascript
if (author.picture)
....
```

**Comment by lightdiscord** [comment]
Maybe close this pr ? I'm not working on it anymore 😢

**Pull request closed without merging**